orttraining/orttraining/models/runner/training_runner.cc

```diff
@@ -634,50 +634,70 @@ void TrainingRunner::RunWithUpdate(VectorString& feed_names,
                                    VectorString& fetch_names,
                                    std::vector<MLValue>& feeds,
                                    std::vector<MLValue>& fetches) {
-  // Cyclically pick up a worker ID.
-  const size_t worker_id = step_ % params_.pipeline_parallel_size;
-
-  // Wait for the previous work to finish its job.
-  // Its resource cannot be overrided when it's still working.
-  pipeline_worker_pool_.Join(worker_id);
-
-  // Copy thread-used variable to thread-specific buffer to maintain their life.
-  pipeline_worker_pool_.worker_states[worker_id].feed_names = feed_names;
-  pipeline_worker_pool_.worker_states[worker_id].feeds = feeds;
-  pipeline_worker_pool_.worker_states[worker_id].fetch_names = fetch_names;
-  pipeline_worker_pool_.worker_states[worker_id].fetches = std::vector<MLValue>();
-
-  Status status = Status::OK();
-  pipeline_worker_pool_.workers[worker_id] = std::thread([&](
-      const size_t worker_id, const size_t step) {
+  if (params_.pipeline_parallel_size > 1) {
+    // Cyclically pick up a worker ID.
+    const size_t worker_id = step_ % params_.pipeline_parallel_size;
+
+    // Wait for the previous work to finish its job.
+    // Its resource cannot be overrided when it's still working.
+    pipeline_worker_pool_.Join(worker_id);
+
+    // Copy thread-used variable to thread-specific buffer to maintain their life.
+    pipeline_worker_pool_.worker_states[worker_id].feed_names = feed_names;
+    pipeline_worker_pool_.worker_states[worker_id].feeds = feeds;
+    pipeline_worker_pool_.worker_states[worker_id].fetch_names = fetch_names;
+    pipeline_worker_pool_.worker_states[worker_id].fetches = std::vector<MLValue>();
+
+    Status status = Status::OK();
+    pipeline_worker_pool_.workers[worker_id] = std::thread([&](
+        const size_t worker_id, const size_t step) {
+#ifdef ENABLE_NVTX_PROFILE
+      // Store the tag for the thread which runs session_.Run(...).
+      // It will be used to name range in Nvidia's visual profiler.
+      auto& profile_context = profile::Context::GetInstance();
+      profile_context.SetThreadTag(
+        std::this_thread::get_id(), std::to_string(step));
+#else
+      ORT_UNUSED_PARAMETER(step);
+#endif
+      status = session_.Run(
```

Comment:
>status = session_.Run [](start = 6, length = 21)

pre-existing: there could be a race condition if all worker threads try to write their own status to this memory. Also, even the worker threads won't execute session_.Run at the same time, If some thread has an error status, it can be overwritten by later worker thread's session_.Run() result. 

Maybe we can use the future object to retrieve the status from each worker? Or we can create a vector of status, and each worker assign it's value on its own slot.

Reply:
I have moved `status` declared to this thread's function; one thread has its own local status and there is no race condition.

---
In reply to: [441871864](https://github.com/microsoft/onnxruntime/pull/4265#discussion_r441871864) [](ancestors = 441871864)

Reply:
Cool. Thanks!

---
In reply to: [441872954](https://github.com/microsoft/onnxruntime/pull/4265#discussion_r441872954) [](ancestors = 441872954,441871864)